src/environment-logs-provider.ts

```diff
@@ -0,0 +1,168 @@
+  }
+
+  get isAborted() {
+    return this._isAborted;
+  }
+
+  private log(value: string) {
+    if (!this.isAborted) {
+      const channel = EnvironmentLogsProvider.environmentLogsOutputChannel;
+      if (!channel) {
+        throw new Error(
+          "environment logs output channel used before initialized"
+        );
+      }
+      channel.appendLine(value);
+    }
+  }
+
+  private async logDeployment(deploymentId: string) {
+    let previousStatus;
+    EnvironmentLogsProvider.environmentLogsOutputChannel.show();
+
+    const deployment = await apiClient.getDeployment(
+      deploymentId,
+      this.abortController
+    );
+
+    if (deployment.status === DeploymentStatus.QUEUED) {
+      this.log(`Deployment is queued! Waiting for it to start...`);
+    }
+
+    while (!this.isAborted) {
+      const { type, status } = await apiClient.getDeployment(
+        deploymentId,
+        this.abortController
+      );
+
+      if (
+        status === DeploymentStatus.QUEUED &&
+        previousStatus === DeploymentStatus.QUEUED
+      ) {
+        this.log(
+          "Queued deployment is still waiting for earlier deployments to finish..."
+        );
+      }
+
+      if (
+        status === DeploymentStatus.IN_PROGRESS &&
+        previousStatus === DeploymentStatus.QUEUED
+      ) {
+        this.log(
+          `Deployment reached its turn! ${
+            type === "deploy" ? "Deploying" : "Destroying"
+          } environment...`
+        );
+      }
+
+      await this.processDeploymentSteps(deploymentId);
+
+      if (
+        ![DeploymentStatus.QUEUED, DeploymentStatus.IN_PROGRESS].includes(
+          status
+        )
+      ) {
+        if (status === "WAITING_FOR_USER") {
+          this.log(
+            "Deployment is waiting for an approval. Run 'env0 approve' or 'env0 cancel' to continue."
+          );
+        }
+
+        return status;
+      }
+
+      previousStatus = status;
+      await sleep(pollStepLogsInterval);
+    }
+  }
+
+  private async processDeploymentSteps(deploymentId: string) {
+    const steps = await apiClient.getDeploymentSteps(
+      deploymentId,
+      this.abortController
+    );
+
+    for (const step of steps) {
+      const alreadyLogged = this.stepsAlreadyLogged.includes(step.name);
+
+      if (!alreadyLogged && step.status !== DeploymentStepStatus.NOT_STARTED) {
+        this.log(`$$$ ${step.name}`);
+        this.log("#".repeat(100));
+        await this.writeDeploymentStepLog(deploymentId, step.name);
+        this.stepsAlreadyLogged.push(step.name);
+      }
+    }
+  }
+
+  private async writeDeploymentStepLog(
+    deploymentLogId: string,
+    stepName: string
+  ) {
+    let shouldPoll = false;
+    let startTime: number | string;
+
+    do {
+      const steps = await apiClient.getDeploymentSteps(
+        deploymentLogId,
+        this.abortController
+      );
+
+      const { status } = steps.find((step) => step.name === stepName) || {};
```

Comment:
why not querying a single step? it seems we have such an endpoint

Reply:
https://github.com/env0/env0/blob/2ceeebb685dd92379d4ec874caabc0ffd6ce64ca/services/deployment-step-service/openapi.json#L162

Reply:
this endpoint return the log not the status